### PR TITLE
Move tools for indexing to TaskToolbox instead of injecting them in constructor

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

This is a simple refactoring to have all tools necessary for indexing in `TaskToolbox` rather than having them separately in different classes. This PR also fixes a bug in `PartialSegmentGenerateTask` which [it always creates `DropwizardRowIngestionMeters` instead of using a configured factory](https://github.com/apache/druid/pull/10308/files#diff-89c55f61b470d7633f336aa0b92f8b49L158) (even though `DropwizardRowIngestionMeters` is the only implementation currently).

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `TaskToolbox`
 * `TaskToolboxFactory`

### CODE REVIEW

**Comment by jihoonson** [comment]
```
411 L | B(0/2)          if (toolbox.getAppenderatorsManager().shouldTaskMakeNodeAnnouncements()) {

898 L                     toolbox.getChatHandlerProvider().unregister(task.getId());

900 L | B(0/2)            if (toolbox.getAppenderatorsManager().shouldTaskMakeNodeAnnouncements()) {
```

The coverage check bot complains about these lines, but [there is no change in them](https://github.com/apache/druid/pull/10308/files#diff-2512ef23844750284130758031054081R411-R900). I think it's a good idea to test node announcement in streaming ingestion with both middleManagers and Indexers, but seems better to do in a separate PR.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java [hunk 1]
```diff
@@ -160,46 +156,33 @@ private static String makeGroupId(boolean isAppendToExisting, String dataSource)
...
-  private final Optional<ChatHandlerProvider> chatHandlerProvider;
+  private final CircularBuffer<Throwable> buildSegmentsSavedParseExceptions;
 
-  @JsonIgnore
-  private final RowIngestionMeters determinePartitionsMeters;
+  private final CircularBuffer<Throwable> determinePartitionsSavedParseExceptions;
 
-  @JsonIgnore
-  private final RowIngestionMeters buildSegmentsMeters;
+  @MonotonicNonNull
```
Comment (jon-wei) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:167]: Could this be `javax.annotation.Nonnull` instead? That seems more common in the codebase, and these objects don't really seem "monotonic"

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialSegmentMergeTaskSuccessReport.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * under the License.
+ */
+
+package org.apache.druid.indexing.common.task.batch.parallel;
+
+import org.apache.druid.timeline.DataSegment;
+
+import java.util.Set;
+
+public class PartialSegmentMergeTaskSuccessReport
```
Comment (jon-wei) [inline indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialSegmentMergeTaskSuccessReport.java:26]: Is this class currently used anywhere?

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/PartialSegmentMergeTaskSuccessReport.java]: Oops, accidentally included.

File: indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java (hunk 1 above)
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:167]: I think `Nonnull` and `MonotonicNonNull` are different. The [javadoc of `Nonnull` says "The annotated element must not be null. Annotated fields must not be null after construction has completed."](https://www.javadoc.io/doc/com.google.code.findbugs/jsr305/latest/javax/annotation/Nonnull.html), while [the one of `MonotonicNonNull` says "Indicates that once the field (or variable) becomes non-null, it never becomes null again."](https://checkerframework.org/api/org/checkerframework/checker/nullness/qual/MonotonicNonNull.html). For the variables annotated as `MonotonicNonNull` in this class, they are nulls when the class is constructed and initialized later in `runTask()`. Once they are initialized, they never become nulls again until the task is finished. So, I think `MonotonicNonNull` is more proper for these variables.

Comment (jon-wei) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/task/IndexTask.java:167]: Ah, I see, that's good then

**Comment by jihoonson** [comment]
@jon-wei thanks for the review!

**Pull request merged**